vpr/test/test_noc_place_utils.cpp

```diff
@@ -0,0 +1,1666 @@
+    delta_laten_cost = 0.;
+
+    // call the test function
+    number_of_affected_traffic_flows = find_affected_noc_routers_and_update_noc_costs(blocks_affected, delta_aggr_band_cost, delta_laten_cost, noc_opts);
+
+    // update the test total noc bandwidth and latency costs based on the cost changes found by the test functions
+    test_noc_bandwidth_costs += delta_aggr_band_cost;
+    test_noc_latency_costs += delta_laten_cost;
+
+    // need this function to update the local datastructures that store all the traffic flow costs
+    commit_noc_costs(number_of_affected_traffic_flows);
+
+    // clear the affected blocks
+    clear_move_blocks(blocks_affected);
+
+    // now verify the test function by comparing the link bandwidths in the noc model (should have been updated by the test function) to the golden set
+    int number_of_links = golden_link_bandwidths.size();
+    for (int link_number = 0; link_number < number_of_links; link_number++) {
+        NocLinkId current_link_id = (NocLinkId)link_number;
+        const NocLink& current_link = noc_ctx.noc_model.get_single_noc_link(current_link_id);
+
+        REQUIRE(golden_link_bandwidths[current_link_id] == current_link.get_bandwidth_usage());
+    }
+
+    // now find the total expected noc aggregate bandwidth and latency cost
+    double golden_total_noc_aggr_bandwidth_cost = 0.;
+    double golden_total_noc_latency_cost = 0.;
+
+    for (int traffic_flow_number = 0; traffic_flow_number < number_of_created_traffic_flows; traffic_flow_number++) {
+        golden_total_noc_aggr_bandwidth_cost += golden_traffic_flow_bandwidth_costs[(NocTrafficFlowId)traffic_flow_number];
+        golden_total_noc_latency_cost += golden_traffic_flow_latency_costs[(NocTrafficFlowId)traffic_flow_number];
+    }
+
+    // now check whether the expected noc costs that we manually calculated above match the noc costs found through the test function (we allow for a tolerance of difference)
+    REQUIRE(vtr::isclose(golden_total_noc_latency_cost, test_noc_latency_costs));
+    REQUIRE(vtr::isclose(golden_total_noc_aggr_bandwidth_cost, test_noc_bandwidth_costs));
+
+    // now test the recompute cost function //
+    // The recompute cost function just adds up all traffic flow costs, so it match the expected noc costs that we manually calculated above by summing up all the expected individual traffic flow costs. //
+
+    // start by resetting the test cost variables
+    test_noc_bandwidth_costs = 0.;
+    test_noc_latency_costs = 0.;
+
+    // now execute the test function
+    recompute_noc_costs(&test_noc_bandwidth_costs, &test_noc_latency_costs);
+
+    // now verify
+    REQUIRE(vtr::isclose(golden_total_noc_latency_cost, test_noc_latency_costs));
+    REQUIRE(vtr::isclose(golden_total_noc_aggr_bandwidth_cost, test_noc_bandwidth_costs));
+
+    // delete local datastructures
+    free_noc_placement_structs();
+
+    // need to delete local noc routing algorithm
+    delete routing_algorithm;
+}
```

Comment:
The above code was a very long test case.  Should it use helper functions?

Reply:
Will leave as is for now. Not critical.